File: playwright-typescript/src/pages/catalog-page.ts
import assert from "node:assert/strict";
import { expect, type Locator, type Page } from "@playwright/test";
import type { ProductRecord } from "../support/models.js";
import { settings } from "../support/settings.js";

export class CatalogPage {
  constructor(private readonly page: Page) {}

  get productCards(): Locator {
    return this.page.locator(".product-card");
  }

  productCard(name: string): Locator {
    return this.productCards.filter({
      has: this.page.getByRole("heading", { name, exact: true })
    });
  }

  async open(): Promise<void> {
    await this.page.goto(settings.baseUrl);
    await this.waitForCatalog();
  }

  async expectOpen(): Promise<void> {
    await expect(this.page).toHaveURL(`${settings.baseUrl}/`);
  }

  async expectHeadingAndCatalog(): Promise<void> {
    await this.waitForCatalog();
    await expect(this.page.getByRole("heading", { name: "Welcome to FictoShop", exact: true })).toBeVisible();
    await expect(this.page.locator("#catalog")).toBeVisible();
    assert.ok((await this.productCards.count()) > 0);
  }

  async expectCardsHavePurchasingInformation(): Promise<void> {
    await this.waitForCatalog();
    const count = await this.productCards.count();
    assert.ok(count > 0);
    for (let index = 0; index < count; index += 1) {
      const card = this.productCards.nth(index);
      await expect(card.locator("h3")).not.toHaveText("");
      await expect(card.locator(".product-price")).toContainText("$");
      await expect(card.locator(".badge")).toBeVisible();
    }
  }

  expectAvailableAddEnabled(name: string): Promise<void> {
    return expect(this.productCard(name).getByRole("button", { name: "Add to cart", exact: true })).toBeEnabled();
  }

  openProduct(name: string): Promise<void> {
    return this.productCard(name).locator(".product-card-title").click();
  }

  async expectProductDetail(product: ProductRecord): Promise<void> {
    await expect(this.page).toHaveURL(new RegExp(`/products/${product.id}/view$`));
    await expect(this.page.getByRole("heading", { name: product.name, exact: true })).toBeVisible();
    await expect(this.page.getByText(product.description, { exact: true })).toBeVisible();
    await expect(this.page.locator(".product-detail-price")).toContainText(`$${product.price.toFixed(2)}`);
    await expect(this.page.locator(".product-detail-rating")).toBeVisible();
    await expect(this.page.locator("#reviews")).toBeVisible();
  }

  search(query: string): Promise<void> {
    return this.page.locator("#product-search").fill(query);
  }

  async expectOnlyProduct(name: string): Promise<void> {
    await expect(this.productCards).toHaveCount(1);
    await expect(this.productCards.first().locator("h3")).toHaveText(name);
  }

  expectCatalogStatus(matches: number, total: number): Promise<void> {
    return expect(this.page.locator("#product-status")).toContainText(`Showing ${matches} of ${total}`);
  }

  async expectNoProducts(): Promise<void> {
    await expect(this.productCards).toHaveCount(0);
    await expect(this.page.locator("#product-empty")).toBeVisible();
  }

  clearSearch(): Promise<void> {
    return this.search("");
  }

  expectProductCount(count: number): Promise<void> {
    return expect(this.productCards).toHaveCount(count);
  }

  selectSort(label: string): Promise<string[]> {
    return this.page.locator("#product-sort").selectOption({ label });
  }

  async displayedPrices(): Promise<number[]> {
    return (await this.productCards.locator(".product-price").allTextContents())
      .map((value) => Number(value.replace("$", "").replaceAll(",", "")));
  }

  async displayedStocks(): Promise<number[]> {
    const result: number[] = [];
    for (let index = 0; index < await this.productCards.count(); index += 1) {
      result.push(Number(await this.productCards.nth(index).locator(".quantity-input").getAttribute("max") ?? 0));
    }
    return result;
  }

  enableInStockOnly(): Promise<void> {
    return this.page.locator("#product-in-stock").check();
  }

  expectProductHidden(name: string): Promise<void> {
    return expect(this.productCard(name)).toHaveCount(0);
  }

  async expectOutOfStockControls(name: string): Promise<void> {
    const card = this.productCard(name);
    await expect(card.locator(".badge")).toHaveText("Out of stock");
    await expect(card.locator(".quantity-input")).toBeDisabled();
    await expect(card.locator(".quantity-btn").first()).toBeDisabled();
    await expect(card.locator(".quantity-btn").last()).toBeDisabled();
    await expect(card.getByRole("button", { name: "Add to cart", exact: true })).toBeDisabled();
  }

  decreaseInitialQuantity(name: string): Promise<void> {
    return this.productCard(name).locator(".quantity-btn").first().click();
  }

  expectQuantity(name: string, quantity: number): Promise<void> {
    return expect(this.productCard(name).locator(".quantity-input")).toHaveValue(String(quantity));
  }

  async increaseBeyondStock(name: string, stock: number): Promise<void> {
    const plus = this.productCard(name).locator(".quantity-btn").last();
    for (let index = 0; index < stock + 2; index += 1) {
      await plus.click();
    }
  }

  exploreApi(): Promise<void> {
    return this.page.getByRole("link", { name: "Explore the API", exact: true }).click();
  }

  expectProductsEndpoint(): Promise<void> {
    return expect(this.page).toHaveURL(`${settings.baseUrl}/products`);
  }

  async expectApiCatalogProducts(): Promise<void> {
    assert.match(await this.page.locator("body").innerText(), /name/);
  }

  expectSignedInUser(username: string): Promise<void> {
    return expect(this.page.locator(".nav-user")).toHaveText(`Signed in as ${username}`);
  }

  expectLogoutButton(): Promise<void> {
    return expect(this.page.locator("header").getByRole("button", { name: "Log out", exact: true })).toBeVisible();
  }

  expectProductVisible(name: string): Promise<void> {
    return expect(this.productCard(name)).toBeVisible();
  }

  expectProductStock(name: string, stock: number): Promise<void> {
    return expect(this.productCard(name).locator(".badge")).toHaveText(`${stock} in stock`);
  }

  async expectProductImage(name: string): Promise<void> {
    const image = this.productCard(name).getByRole("img", { name: `${name} photo` });
    await image.scrollIntoViewIfNeeded();
    await expect(image).toBeVisible();
    await expect(image).toHaveAttribute("src", /\/images\/uploads\//);
    await expect(image).toHaveJSProperty("complete", true, { timeout: 15_000 });
    assert.equal(await image.evaluate((element: HTMLImageElement) => element.naturalWidth > 0), true);
  }

  private waitForCatalog(): Promise<void> {
    return expect(this.page.locator("#product-grid")).toHaveAttribute("aria-busy", "false");
  }
}

File: playwright-csharp/Pages/CatalogPage.cs
using System.Globalization;
using System.Text.RegularExpressions;
using Fictoshop.PlaywrightTests.Support;
using Microsoft.Playwright;
using NUnit.Framework;
using static Microsoft.Playwright.Assertions;

namespace Fictoshop.PlaywrightTests.Pages;

public sealed class CatalogPage
{
    private readonly IPage _page;

    public CatalogPage(IPage page)
    {
        _page = page;
    }

    public ILocator ProductCards => _page.Locator(".product-card");

    public ILocator ProductCard(string name) => ProductCards.Filter(new LocatorFilterOptions
    {
        Has = _page.GetByRole(AriaRole.Heading, new() { Name = name, Exact = true }),
    });

    public async Task OpenAsync()
    {
        await _page.GotoAsync(TestSettings.BaseUrl);
        await WaitForCatalogAsync();
    }

    public async Task AssertOpenAsync() => await Expect(_page).ToHaveURLAsync(TestSettings.BaseUrl + "/");

    public async Task AssertHeadingAndCatalogAsync()
    {
        await WaitForCatalogAsync();
        await Expect(_page.GetByRole(AriaRole.Heading, new() { Name = "Welcome to FictoShop", Exact = true })).ToBeVisibleAsync();
        await Expect(_page.Locator("#catalog")).ToBeVisibleAsync();
        Assert.That(await ProductCards.CountAsync(), Is.GreaterThan(0));
    }

    public async Task AssertCardsHavePurchasingInformationAsync()
    {
        await WaitForCatalogAsync();
        var count = await ProductCards.CountAsync();
        Assert.That(count, Is.GreaterThan(0));
        for (var index = 0; index < count; index++)
        {
            var card = ProductCards.Nth(index);
            await Expect(card.Locator("h3")).Not.ToHaveTextAsync("");
            await Expect(card.Locator(".product-price")).ToContainTextAsync("$");
            await Expect(card.Locator(".badge")).ToBeVisibleAsync();
        }
    }

    public Task AssertAvailableAddEnabledAsync(string name) =>
        Expect(ProductCard(name).GetByRole(AriaRole.Button, new() { Name = "Add to cart", Exact = true })).ToBeEnabledAsync();

    public Task OpenProductAsync(string name) => ProductCard(name).Locator(".product-card-title").ClickAsync();

    public async Task AssertProductDetailAsync(ProductRecord product)
    {
        await Expect(_page).ToHaveURLAsync(new Regex($@"/products/{product.Id}/view$"));
        await Expect(_page.GetByRole(AriaRole.Heading, new() { Name = product.Name, Exact = true })).ToBeVisibleAsync();
        await Expect(_page.GetByText(product.Description, new() { Exact = true })).ToBeVisibleAsync();
        await Expect(_page.Locator(".product-detail-price")).ToContainTextAsync($"${product.Price:F2}");
        await Expect(_page.Locator(".product-detail-rating")).ToBeVisibleAsync();
        await Expect(_page.Locator("#reviews")).ToBeVisibleAsync();
    }

    public Task SearchAsync(string query) => _page.Locator("#product-search").FillAsync(query);

    public async Task AssertOnlyProductAsync(string name)
    {
        await Expect(ProductCards).ToHaveCountAsync(1);
        await Expect(ProductCards.First.Locator("h3")).ToHaveTextAsync(name);
    }

    public Task AssertCatalogStatusAsync(int matches, int total) =>
        Expect(_page.Locator("#product-status")).ToContainTextAsync($"Showing {matches} of {total}");

    public async Task AssertNoProductsAsync()
    {
        await Expect(ProductCards).ToHaveCountAsync(0);
        await Expect(_page.Locator("#product-empty")).ToBeVisibleAsync();
    }

    public Task ClearSearchAsync() => SearchAsync("");
    public Task AssertProductCountAsync(int count) => Expect(ProductCards).ToHaveCountAsync(count);
    public Task SelectSortAsync(string label) => _page.Locator("#product-sort").SelectOptionAsync(new SelectOptionValue { Label = label });

    public async Task<IReadOnlyList<decimal>> DisplayedPricesAsync()
    {
        var values = await ProductCards.Locator(".product-price").AllTextContentsAsync();
        return values.Select(value => decimal.Parse(value.Replace("$", "").Replace(",", ""), CultureInfo.InvariantCulture)).ToList();
    }

    public async Task<IReadOnlyList<int>> DisplayedStocksAsync()
    {
        var result = new List<int>();
        for (var index = 0; index < await ProductCards.CountAsync(); index++)
        {
            var value = await ProductCards.Nth(index).Locator(".quantity-input").GetAttributeAsync("max");
            result.Add(int.Parse(value ?? "0", CultureInfo.InvariantCulture));
        }

        return result;
    }

    public Task EnableInStockOnlyAsync() => _page.Locator("#product-in-stock").CheckAsync();
    public Task AssertProductHiddenAsync(string name) => Expect(ProductCard(name)).ToHaveCountAsync(0);

    public async Task AssertOutOfStockControlsAsync(string name)
    {
        var card = ProductCard(name);
        await Expect(card.Locator(".badge")).ToHaveTextAsync("Out of stock");
        await Expect(card.Locator(".quantity-input")).ToBeDisabledAsync();
        await Expect(card.Locator(".quantity-btn").First).ToBeDisabledAsync();
        await Expect(card.Locator(".quantity-btn").Last).ToBeDisabledAsync();
        await Expect(card.GetByRole(AriaRole.Button, new() { Name = "Add to cart", Exact = true })).ToBeDisabledAsync();
    }

    public Task DecreaseInitialQuantityAsync(string name) => ProductCard(name).Locator(".quantity-btn").First.ClickAsync();
    public Task AssertQuantityAsync(string name, int quantity) => Expect(ProductCard(name).Locator(".quantity-input")).ToHaveValueAsync(quantity.ToString());

    public async Task IncreaseBeyondStockAsync(string name, int stock)
    {
        var plus = ProductCard(name).Locator(".quantity-btn").Last;
        for (var index = 0; index < stock + 2; index++)
        {
            await plus.ClickAsync();
        }
    }

    public Task ExploreApiAsync() => _page.GetByRole(AriaRole.Link, new() { Name = "Explore the API", Exact = true }).ClickAsync();
    public Task AssertProductsEndpointAsync() => Expect(_page).ToHaveURLAsync(TestSettings.BaseUrl + "/products");

    public async Task AssertApiCatalogProductsAsync()
    {
        var body = await _page.Locator("body").InnerTextAsync();
        Assert.That(body, Does.Contain("name"));
    }

    public Task AssertSignedInUserAsync(string username) => Expect(_page.Locator(".nav-user")).ToHaveTextAsync($"Signed in as {username}");
    public Task AssertLogoutButtonAsync() => Expect(_page.Locator("header").GetByRole(AriaRole.Button, new() { Name = "Log out", Exact = true })).ToBeVisibleAsync();
    public Task AssertProductVisibleAsync(string name) => Expect(ProductCard(name)).ToBeVisibleAsync();
    public Task AssertProductStockAsync(string name, int stock) => Expect(ProductCard(name).Locator(".badge")).ToHaveTextAsync($"{stock} in stock");

    public async Task AssertProductImageAsync(string name)
    {
        var image = ProductCard(name).GetByRole(AriaRole.Img, new() { Name = $"{name} photo" });
        await image.ScrollIntoViewIfNeededAsync();
        await Expect(image).ToBeVisibleAsync();
        await Expect(image).ToHaveAttributeAsync("src", new Regex("/images/uploads/"));
        await Expect(image).ToHaveJSPropertyAsync("complete", true, new() { Timeout = 15_000 });
        Assert.That(await image.EvaluateAsync<bool>("element => element.naturalWidth > 0"), Is.True);
    }

    private Task WaitForCatalogAsync() => Expect(_page.Locator("#product-grid")).ToHaveAttributeAsync("aria-busy", "false");
}
